File: Data Structures & Algorithms/binary-tree-diameter/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def height(node):
            if not node:
                return 0
            return 1 + max(height(node.left), height(node.right))
        if not root:
            return 0
        if not root.left or not root.right:
            if not root.left:
                return max(height(root.left) + height(root.right), self.diameterOfBinaryTree(root.right))
            else:
                return max(height(root.left) + height(root.right), self.diameterOfBinaryTree(root.left))

        return height(root.left) + height(root.right)

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* dfs(TreeNode* node, int p, int q) {
        if (node->val >= p and node->val <= q) {
            return node;
        }

        if (node->val > p and node->val > q) {
            return dfs(node->left, p, q);
        } else {
            return dfs(node->right, p, q);
        }


    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return dfs(root, min(p->val, q->val), max(p->val, q->val));
    }
};
